File: day_06/main.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

func loadRawFile(fileName string) string {
	data, err := os.ReadFile(fileName)
	if err != nil {
		panic(err)
	}
	return strings.TrimSpace(string(data))
}

func removeEmptyLines(input string) [][]string {
	lines := strings.Split(input, "\n")
	var result [][]string
	for _, line := range lines {
		trimmedLine := strings.TrimSpace(line)
		if trimmedLine != "" {
			result = append(result, strings.Split(trimmedLine, ""))
		}
	}
	return result
}

type Position struct {
	X, Y int
}

func getGuardPosition(data [][]string) Position {
	for y, row := range data {
		for x, tile := range row {
			if tile == "^" {
				return Position{X: x, Y: y}
			}
		}
	}
	panic("Guard position not found")
}

func isWithinBounds(data [][]string, position Position) bool {
	return position.Y >= 0 && position.Y < len(data) && position.X >= 0 && position.X < len(data[position.Y])
}

func turnRight(direction Position) Position {
	return Position{X: -direction.Y, Y: direction.X}
}

func executePart1(input string) int {
	data := removeEmptyLines(input)
	guardPosition := getGuardPosition(data)
	direction := Position{X: 0, Y: -1}
	walkedTiles := 1

	for isWithinBounds(data, guardPosition) {
		nextPosition := Position{X: guardPosition.X + direction.X, Y: guardPosition.Y + direction.Y}

		if !isWithinBounds(data, nextPosition) {
			return walkedTiles
		}

		nextTile := data[nextPosition.Y][nextPosition.X]
		if nextTile == "#" {
			direction = turnRight(direction)
		} else {
			if data[guardPosition.Y][guardPosition.X] != "X" {
				walkedTiles++
			}
			data[guardPosition.Y][guardPosition.X] = "X"
			guardPosition = nextPosition
		}
	}
	return walkedTiles
}

func canExitMaze(data [][]string, guardPosition Position, guardDirection Position) bool {
	position := guardPosition
	direction := guardDirection
	turns := make(map[string]bool)

	for isWithinBounds(data, position) {
		nextPosition := Position{X: position.X + direction.X, Y: position.Y + direction.Y}

		if !isWithinBounds(data, nextPosition) {
			return true
		}

		nextTile := data[nextPosition.Y][nextPosition.X]
		if nextTile == "#" || nextTile == "O" {
			command := fmt.Sprintf("%d,%d:%d,%d", position.X, position.Y, direction.X, direction.Y)
			direction = turnRight(direction)
			if turns[command] {
				return false
			}
			turns[command] = true
		} else {
			data[position.Y][position.X] = "X"
			position = nextPosition
		}
	}
	return false
}

func executePart2(input string) int {
	data := removeEmptyLines(input)
	initialPosition := getGuardPosition(data)
	position := initialPosition
	direction := Position{X: 0, Y: -1}
	successfulChanges := make(map[string]bool)

	for isWithinBounds(data, position) {
		nextPosition := Position{X: position.X + direction.X, Y: position.Y + direction.Y}

		if !isWithinBounds(data, nextPosition) {
			break
		}

		nextTile := data[nextPosition.Y][nextPosition.X]
		if nextTile == "#" || nextTile == "O" {
			direction = turnRight(direction)
		} else {
			nextData := make([][]string, len(data))
			for i := range data {
				nextData[i] = append([]string{}, data[i]...)
			}
			nextData[nextPosition.Y][nextPosition.X] = "O"
			if !canExitMaze(nextData, initialPosition, Position{X: 0, Y: -1}) {
				key := fmt.Sprintf("%d,%d", nextPosition.X, nextPosition.Y)
				successfulChanges[key] = true
			}
			position = nextPosition
		}
	}

	return len(successfulChanges)
}

func main() {
	inputPath := filepath.Join(".", "input.txt")
	input := loadRawFile(inputPath)

	fmt.Println("Part 1:", executePart1(input))
	fmt.Println("Part 2:", executePart2(input))
}

File: day_06/main.py
import os

def load_raw_file(file_name):
    with open(file_name, 'r') as file:
        return file.read().strip()

def remove_empty_lines(input):
    return [list(line) for line in input.split("\n") if line.strip()]

def get_guard_position(data):
    for y, row in enumerate(data):
        if "^" in row:
            x = row.index("^")
            return {"x": x, "y": y}
    raise ValueError("Guard position not found")

def is_within_bounds(data, position):
    return (
        0 <= position["y"] < len(data) and
        0 <= position["x"] < len(data[position["y"]])
    )

def turn_right(direction):
    return {"x": -direction["y"], "y": direction["x"]}

def execute_part1(input):
    data = remove_empty_lines(input)
    guard_position = get_guard_position(data)
    direction = {"x": 0, "y": -1}
    walked_tiles = 1

    while is_within_bounds(data, guard_position):
        next_position = {
            "x": guard_position["x"] + direction["x"],
            "y": guard_position["y"] + direction["y"]
        }

        if not is_within_bounds(data, next_position):
            return walked_tiles

        next_tile = data[next_position["y"]][next_position["x"]]
        if next_tile == "#":
            direction = turn_right(direction)
        else:
            if data[guard_position["y"]][guard_position["x"]] != "X":
                walked_tiles += 1
            data[guard_position["y"]][guard_position["x"]] = "X"
            guard_position = next_position

    return walked_tiles

def can_exit_maze(data, guard_position, guard_direction):
    position = guard_position.copy()
    direction = guard_direction.copy()
    turns = set()

    while is_within_bounds(data, position):
        next_position = {
            "x": position["x"] + direction["x"],
            "y": position["y"] + direction["y"]
        }

        if not is_within_bounds(data, next_position):
            return True

        next_tile = data[next_position["y"]][next_position["x"]]
        if next_tile in ["#", "O"]:
            command = f'{position["x"]},{position["y"]}:{direction["x"]},{direction["y"]}'
            direction = turn_right(direction)
            if command in turns:
                return False
            turns.add(command)
        else:
            data[position["y"]][position["x"]] = "X"
            position = next_position

    return False

def execute_part2(input):
    data = remove_empty_lines(input)
    initial_position = get_guard_position(data)
    position = initial_position.copy()
    direction = {"x": 0, "y": -1}
    successful_changes = set()

    while is_within_bounds(data, position):
        next_position = {
            "x": position["x"] + direction["x"],
            "y": position["y"] + direction["y"]
        }

        if not is_within_bounds(data, next_position):
            break

        next_tile = data[next_position["y"]][next_position["x"]]
        if next_tile in ["#", "O"]:
            direction = turn_right(direction)
        else:
            next_data = [row[:] for row in data]
            next_data[next_position["y"]][next_position["x"]] = "O"
            if not can_exit_maze(next_data, initial_position, {"x": 0, "y": -1}):
                successful_changes.add(f'{next_position["x"]},{next_position["y"]}')
            position = next_position

    return len(successful_changes)

if __name__ == "__main__":
    input_path = os.path.join(".", "input.txt")
    input = load_raw_file(input_path)

    example_input = """....#.....
.........#
..........
..#.......
.......#..
..........
.#..^.....
........#.
#.........
......#..."""

    print("Part 1:", execute_part1(input))
    print("Part 2:", execute_part2(input))

